File: dp_educative/1_knapsack/01-knapsack.py

from functools import cache, lru_cache


def solve_knapsack(profits, weights, capacity):
    def knapsack_recursive(profits, weights, capacity, currIndex=0):
        if capacity <= 0 or currIndex > len(profits)-1:
            return 0

        profit1 = profit2 = 0
        if weights[currIndex] <= capacity:
            profit1 = profits[currIndex] + knapsack_recursive(
                profits, weights, capacity-weights[currIndex], currIndex+1)

        profit2 = knapsack_recursive(profits, weights, capacity, currIndex+1)

        return max(profit1, profit2)

    return knapsack_recursive(profits, weights, capacity)


print(solve_knapsack([15, 20, 50], [1, 2, 3], 5))
print(solve_knapsack([1, 6, 10, 16], [1, 2, 3, 5], 8))
print(solve_knapsack([1, 6, 10, 16], [1, 2, 3, 5], 6))

File: dp_educative/1_knapsack/01-knapsack.cpp
#include "bits/stdc++.h"

using namespace std;

class Knapsack {
public:
  int solveKnapsack(const vector<int> &profits, const vector<int> &weights,
                    int capacity) {

    return solveKnapsackRecursion(profits, weights, capacity);
  }

  int solveKnapsackRecursion(const vector<int> &profits,
                             const vector<int> &weights, int capacity,
                             int index = 0) {
    if (index > profits.size() - 1 || capacity <= 0) {
      return 0;
    }
    int p1 = 0, p2 = 0;
    if (capacity >= weights[index]) {
      p1 = profits[index] + solveKnapsackRecursion(profits, weights,
                                                   capacity - weights[index],
                                                   index + 1);
    }

    p2 = solveKnapsackRecursion(profits, weights, capacity, index + 1);
    return max(p1, p2);
  }
};

int main(int argc, char *argv[]) {
  Knapsack ks;
  vector<int> profits = {1, 6, 10, 16};
  vector<int> weights = {1, 2, 3, 5};
  int maxProfit = ks.solveKnapsack(profits, weights, 7);
  cout << "Total knapsack profit ---> " << maxProfit << endl;
  maxProfit = ks.solveKnapsack(profits, weights, 6);
  cout << "Total knapsack profit ---> " << maxProfit << endl;
}